### ISSUE TITLE
Implement hyper modern python

### ISSUE THREAD

**Issue by felnne** [mapaction/rolling-data-scramble-dashboard-poc#50]
From https://cjolowicz.github.io/posts/hypermodern-python-01-setup/ (and follow on posts in series)

- [x] pyenv

initially skipped but completed in #56.

- [x] poetry

I used this already.

- [x] `__version__` variable

Changed to using conventional variable name but not yet reading from package metadata (i.e. `pyproject.toml`)

- [x] package in `src`

Changed to using recommended package layout to avoid differences in how things like imports work between having the code installed as a package vs being in your current directory. See [this blog](https://blog.ionelmc.ro/2014/05/25/python-packaging/) for more information & background.

- [x] Poetry scripts

See #51.

- [x] Pytest

Initially with a stub test. See #55 for task to add real tests.

- [x] Coverage

Set to fail on less than 100%, which will be the case until #55 is complete.

- PyTest mock

Not implemented yet, but will for #55.

- Nox

Postponed until #56

- [x] Flake8

Added, with plugins for:

* Annotations (stupid rule for type hinting `self` disabled (`ANN101`))
* Black (incompatible Flake8 rules disabled (`E203`,`W503`)
* Bugbear (both standard (`B`) and opinionated (`B9`) rules, incompatible Flake8 rules disabled (`E501`)
* Doc blocks (using pep257/sphinx style, some rules disabled:
  *  (`D100`/`D104`: block needed at the front of each module/package)
  * (`D105`: block needed for magic methods (agreed where methods are non-trivial))
* Import ordering (with Google style) - it's noted in the guide there are other options for this as well [1](https://github.com/gforcada/flake8-isort), [2](asottile/reorder-python-imports), [3](https://github.com/sqlalchemyorg/zimports) which I haven't tried yet
* Bandit (security scanning, with use of assert error ignored for tests)

Max line length increased to 120.

- [x] Black

Already added.

- [x] Safety

Initially skipped, now added.

- [x] pre-commit hook

Not initially implemented (https://github.com/mapaction/rolling-data-scramble-dashboard-poc/issues/50#issuecomment-787122818) but now added (https://github.com/mapaction/rolling-data-scramble-dashboard-poc/issues/50#issuecomment-792373548)

- [x] MyPy

This needed some changes to variable names (inadvertently redeclaring the same variable name in loops for example).

I enabled some additional configuration options:

```ini
[mypy]
warn_redundant_casts = True

[mypy-mapaction_rds_dashboard.*]
disallow_untyped_calls = True
disallow_untyped_defs = True
disallow_incomplete_defs = True
check_untyped_defs = True
disallow_untyped_decorators = True

no_implicit_optional = True

warn_unused_ignores = True
```

I think there is still a task to check we only define type hints when needed (#59).

I used the approach from https://github.com/mapaction/rolling-data-scramble-dashboard-poc/issues/50#issuecomment-787436455 a couple of times.

- PyType

Skipped as this doesn't support Python 3.9.

- Desert and Marshmallow

I want to look at this in #30, as I think that approach would work really well. 

It may also be feasible to define a model/schema for the MapChef layer file - though as that has a JSON Schema we could use something [Pydantic](https://pydantic-docs.helpmanual.io/datamodel_code_generator/) to do that (or something less specific that takes a JSON Schema and generates a Data Class rather than a Pydantic model).

- [x] DocStrings

Already added (using the Sphinx format that PyCharm uses by default).

- xdoctest

This felt redundant.

- Sphinx

I don't think this is needed for this project, but it definitely would for others (MapChef for example). I might come back to this later.

- [x] GitHub actions for testing

I adapted the examples given as we don't use Nox yet, I also added a parallel Windows job - I think these could be merged but this might make them harder for people (read: me) to understand.

- [x] importlib (version single source of truth)

Implements part of #8.

- [x] GitHub actions for building package

This was pretty simple but I'm not yet happy with how this works:

* [x] it should follow tests (so workflows need collapsing together)
* [x] it should use a build matrix given the differences between tasks are minor
* [x] it should happen when a pull request is merged (tests should happen on each commit)
* [x] it should verify the package can be installed
* [x] it should push the package to PyPi (waiting on recommendation for how to do this for MapAction) #69

* Read The Docs

As Sphinx was skipped this was skipped too.
